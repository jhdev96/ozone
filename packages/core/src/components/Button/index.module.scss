@import '../../scss/root';
@import '../../scss/mixins/button/button.scss';
@import '../../scss/mixins/button/variables.scss';
                                  

// For styling `flat` buttons
$flatMap: (
  primary: (
    color: #fff,
    bg-color: var(--blue-3)
  ),
  secondary: (
    color: #fff,
    bg-color: var(--red-3)
  ),
  warning: (
    color: var(--black-core),
    bg-color: var(--yellow-3)
  )
);

/* Major button classes - primary, secondary, warning, basic */                                   
.primary {
  @include button;

  color: #fff;
  background-color: var(--blue-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;

    color: var(--blue-core);
    border: 1px solid var(--blue-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--red-core) auto 1px !important;
  }
}

.secondary {
  @include button;

  color: #fff;
  background-image: $secondary-gradient;
  background-repeat: repeat-x;
  border-color: $button-bc1 $button-bc2 $button-bc3 !important;
  box-shadow: inset 0 1px 0 $primary-shade1, 
                    0 1px 2px $primary-shade2;
  background-color: var(--red-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;
    color: var(--red-core);
    border: 1px solid var(--red-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--blue-core) auto 1px !important;
  }
}

.warning {
  @include button;

  color: var(--black-core);
  background-color: var(--yellow-core);
  border-radius: 4px;

  &:focus {
    outline: var(--black-core) auto 1px !important;
  }
}

.basic {
  @include button;

  border-radius: 3px;

  &:focus {
    outline: var(--gray-4) auto 1px !important;
  }
}

/* The following styles serve as modifiers 
 * for the major button classes 
*/
.flat {
  background-image: none !important;
  background-repeat: no-repeat;
  border: 0;
  border-radius: 0 !important;
  box-shadow: none !important;
  text-shadow: none !important;

  @each $class, $rule in $flatMap {
    &.#{$class}:not(.disabled):not(.loading) {
      &:hover {
        color: map-get($rule, 'color');
        background-color: map-get($rule, 'bg-color');
      }
    }
  }
}

.fill { width: 100%; }

.loading {
  padding: 8px 20px 5px 20px;
  max-height: 35px;
  opacity: 0.90;
  cursor: default;

  &:focus {outline: none !important;}
}

.disabled {
  opacity: 0.70;
  cursor: default;
}

.small {
  min-width: fit-content;
  padding: 8px !important;
}

.medium { min-width: 40px; }

.large {
  min-width: 70px;
  max-height: 70px !important;
  padding: 15px 20px !important;
  font-size: 1.2em !important;
}

.light {
  &.basic {
    color: var(--black-core);
    background-color: #fff;
    border: 1px solid var(--gray-3);
  }
}

.dark {
  &.basic {
    color: var(--gray-1);
    border: 1px solid var(--black-6);
    background-color: var(--black-4) !important;
  }
}