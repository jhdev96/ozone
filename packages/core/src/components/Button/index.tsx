import React from 'react';
import classNames from '../../utils/classNames';
import { 
  ButtonVariation,
  loaderTheme
} from './utils';
import LoadingDots from '../LoadingDots';
import styles from './index.module.scss';


export interface ButtonProps {
  /** 
   * The main classes of buttons. 
   */
  variation?: ButtonVariation;
  /** 
   * Sets the HTML `type` attribute for the
   * button element.
   */ 
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"]
  /** 
   * Control the size of the button. Default is medium. 
   */
  size?: "small" | "medium" | "large";
  /** 
   * Give the button a flat texture, appropriate for forms. 
   */
  flat?: boolean;
  /** 
   * When true the button will no background; 
   * It will only have a border. 
   */
  outline?: boolean;
  /** 
   * When true, the button's width will be 100% of 
   * its container. 
   */ 
  fill?: boolean;
  /** 
   * Color the button based on the current theme. 
   */
  theme?: "light" | "dark";
  /** 
   * Make the button disabled. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates whether the button is loading some content. 
   */
  isLoading?: boolean;
  /** 
   * Sets the `className` attribute on the button element. 
   */ 
  className?: string;
  /** 
   * The function that handles click events. 
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers over the button.
   */
  onMouseEnter?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers away from the
   * button.
   */
  onMouseLeave?: React.MouseEventHandler<HTMLButtonElement>;
  /** 
   * A selector used for testing this component. 
   */
  dataTestId?: string;
  /** 
   * Content displayed inside the button. 
   */
  children?: React.ReactNode | string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button({
    variation     = "primary",
    type          = "button",
    size          = "medium", 
    flat          = false,
    outline       = false,
    fill          = false, 
    theme         = "light",
    isDisabled    = false,
    isLoading     = false,
    className,
    onClick       = () => {},
    onMouseEnter  = () => {},
    onMouseLeave  = () => {},
    dataTestId,
    children}: ButtonProps,
    ref: React.Ref<HTMLButtonElement>): JSX.Element 
  {
    const childMarkup = isLoading 
      ? (
          <>
            <LoadingDots 
              dotSize={7} 
              position="center" 
              theme={loaderTheme(variation, theme, outline)} 
            />
            <span style={{opacity: '0'}}>{children}</span>
          </>
        )
      : children;

    return (
      <button
        type={type}
        className={classNames(
          className,
          styles[variation],
          styles[size],
          styles[theme],
          flat        && styles.flat,
          outline     && styles[`${variation}--outline`],
          fill        && styles.fill,
          isLoading   && styles.loading,
          isDisabled  && styles.disabled
        )}
        onClick={isDisabled ? () => {} : onClick}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        disabled={isDisabled}
        aria-busy={isLoading}
        data-testid={dataTestId}
        ref={ref}
      >
        {childMarkup}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;