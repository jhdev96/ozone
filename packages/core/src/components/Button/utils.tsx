import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}