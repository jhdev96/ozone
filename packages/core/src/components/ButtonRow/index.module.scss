$placementMap: (
  'left': flex-start,
  'center': center,
  'right': flex-end
);

.buttonRow {
  display: inline-flex;
  flex-direction: column;
  width: 100%;
}

@each $placement, $rule in $placementMap {
  .#{$placement} { align-items: $rule; }
}