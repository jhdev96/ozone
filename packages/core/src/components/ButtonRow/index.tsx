import React from 'react';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export interface ButtonRowProps {
  /** 
   * The placement of the button(s) in the container. 
   */
  placement?: "left" | "center" | "right";
  /** 
   * The buttons that occupy the row. 
   */
  children?: React.ReactNode;
}

export default function ButtonRow({
  placement = "center", 
  children
}: ButtonRowProps): JSX.Element 
{
  return (
    <div 
      className={classNames(
        styles.buttonRow,
        styles[placement]
      )}
    >
      <div>{children}</div>
    </div>
  );
}

ButtonRow.displayName = 'ButtonRow';