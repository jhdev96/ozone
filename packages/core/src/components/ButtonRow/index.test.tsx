import React from 'react';
import { render } from '@testing-library/react';
import ButtonRow from '.';
import Button from '../Button';


describe('<ButtonRow />', () => {
  it('renders a button row with `left` placement', () => {
    const {container} = render(
      <ButtonRow placement="left">
        <Button variation="basic">Back</Button>
        <Button variation="primary">Continue</Button>
      </ButtonRow>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a button row with `center` placement', () => {
    const {container} = render(
      <ButtonRow placement="center">
        <Button variation="basic">Back</Button>
        <Button variation="primary">Continue</Button>
      </ButtonRow>
    );

    expect(container).toMatchSnapshot();
  }); 
  
  it('renders a button row with `right` placement', () => {
    const {container} = render(
      <ButtonRow placement="left">
        <Button variation="basic">Back</Button>
        <Button variation="primary">Continue</Button>
      </ButtonRow>
    );

    expect(container).toMatchSnapshot();
  });
});