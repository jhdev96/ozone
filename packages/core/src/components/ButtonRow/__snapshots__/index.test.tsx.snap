// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonRow /> renders a button row with \`center\` placement 1`] = `
<div>
  <div
    class="buttonRow center"
  >
    <div>
      <button
        aria-busy="false"
        class="basic medium light"
        type="button"
      >
        Back
      </button>
      <button
        aria-busy="false"
        class="primary medium light"
        type="button"
      >
        Continue
      </button>
    </div>
  </div>
</div>
`;

exports[`<ButtonRow /> renders a button row with \`left\` placement 1`] = `
<div>
  <div
    class="buttonRow left"
  >
    <div>
      <button
        aria-busy="false"
        class="basic medium light"
        type="button"
      >
        Back
      </button>
      <button
        aria-busy="false"
        class="primary medium light"
        type="button"
      >
        Continue
      </button>
    </div>
  </div>
</div>
`;

exports[`<ButtonRow /> renders a button row with \`right\` placement 1`] = `
<div>
  <div
    class="buttonRow left"
  >
    <div>
      <button
        aria-busy="false"
        class="basic medium light"
        type="button"
      >
        Back
      </button>
      <button
        aria-busy="false"
        class="primary medium light"
        type="button"
      >
        Continue
      </button>
    </div>
  </div>
</div>
`;
